README.md
## array-correl

### Installation

`npm install array-correl`

### Generate

The **generate** method generates an array of length **count** with correlated sets of numbers in the `{ x: xValue, y: yValue }` format. Optionally, the following parameters can be specified:

- Target **correlation** coefficient ranging from 0 to 1,
- Target **mean**,
- Target **deviation**

If the correlation coefficient is not specified, it defaults to 0.7; if mean is not specified, it defaults to 0; if deviation is not specified, it defaults to 1.

#### API

array-correl.**generate**(count, [correlation, [mean, [deviation]]]);

#### Errors

If no **count** argument is passed, a **ReferenceError** is thrown. If any passed argument is **not a number** (or cannot be coerced to a number), a **TypeError** is thrown.

### Inspect

The **inspect** method analyzes the supplied array and returns several statistics measures. The format of the input array must be one of the following:

- Each array element contains an object with the following properties: `{ x: xValue, y: yValue }`
- Each array element contains a two-element array of the following structure: `[ xValue, yValue ]`

The function returns the following parameters in a JS object:

- The Pearson correlation coefficient of the x and y data series
- The deviation of each of the the x and y data series
- The mean of each of the x and y data series
- The extent (min and max) of each of the x and y data series

#### API

array-correl.**inspect**(array);

#### Errors

If no **array** argument is passed, a **ReferenceError** is thrown. If the passed array doesn't contain either two-element arrays or objects with x and y properties, a **TypeError** is thrown.

Example:
```
// Import methods and version from array-correl from node_modules
const { generate, inspect, version } = require('array-correl');

// Import methods and version from array-correl locally
// const { generate, inspect, version } = require('../src/index.js');

// Log version
console.log(`array-correl version: ${version}`);

// Define inputs to the generate method
const count = 10000;
const correlation = 0.8;
const mean = 100;
const deviation = 1;

// Log inputs
console.log('\nInputs to generate:');
console.log(`  count: ${count}`);
console.log(`  correlation: ${correlation}`);
console.log(`  mean: ${mean}`);
console.log(`  deviation: ${deviation}`);

// Create an array with values of correlated pairs of numbers
const array = generate(count, correlation, mean, deviation);

console.log('\nFirst 10 correlated pairs:');
// Log the first 10 array elements
for (let i = 0; i < 10; i++) {
  console.log(`  ${JSON.stringify(array[i])}`);
}

// Inspect the array
const result = inspect(array);
console.log(`\nOutput of inspect: \n${JSON.stringify(result, null, 2)}\n`);

/*
Example output:
  array-correl version: 1.0.0

  Inputs to generate:
    count: 10000
    correlation: 0.8
    mean: 100
    deviation: 1

  First 10 correlated pairs:
    {"x":99.71068933565104,"y":99.86127640311815}
    {"x":100.83597240809056,"y":100.074669667149}
    {"x":102.0130498785201,"y":101.38201755974566}
    {"x":100.65061261210614,"y":99.88629765791357}
    {"x":98.30631250480867,"y":98.49421518032264}
    {"x":99.48427439034523,"y":99.81419151256067}
    {"x":98.90338120221003,"y":98.38761587336445}
    {"x":101.06519781042032,"y":101.50229105410696}
    {"x":98.7248721438115,"y":98.06148966509333}
    {"x":99.40834127473076,"y":98.90308575665651}

  Output of inspect:
  {
    "r": 0.802558017048035,
    "xDeviation": 1.0004768422899535,
    "yDeviation": 0.9984232776109369,
    "xExtent": [
      95.99318668690016,
      103.92406651194449
    ],
    "yExtent": [
      96.15152222172428,
      103.84111296300848
    ],
    "xMean": 100.00871360178544,
    "yMean": 100.00635903137874
  }
*/
```


CHANGELOG.md
1.0.1 – 2020-06-14 -
- Entrypoint

1.0.0 – 2020-06-14 – 97d9844c06a8854818d11dc16cc54d7db258b45f
- Added node.js and html5 example uses
- Added local computation of normal distribution
- Removed d3
- Added d3-array
- Added tests, mocha, chai
- Make available in browser
- Lint
- Removed top level index.js

0.1.1 – 2019-03-04 – 8e5ef2b17d8e89676fb30ec937a0ad3c8c227f0e
- Further cleanup

0.0.4 – 2018-04-28 – 7afbc3633b1a920259945496f36861d3d3d8d2eb
- Further cleanup

0.0.3 – 2018-04-28 – 6cc43427f90fe7d35c9258db88ecdffff7b0d1a9
- Cleanup

0.0.2 – 2017-06-04 – 7c37d562d353949f46c586493ca2609215a7112f
- Added keywords, readme edit

0.0.1 – 2017-06-04 – 2b935bc309b7893d51164f102c25c1eef418cb33
- Initial code


test/test.js
/* eslint-disable no-console */
/* global assert, before, describe, expect, it */

const { generate, inspect } = require('../src/index.js');

function approximatelyEqual(value, expected, tolerance) {
  // console.log(value, expected, expected - tolerance, expected + tolerance);
  return (value >= (expected - tolerance)) && (value <= (expected + tolerance));
}

describe('Generate', () => {
  describe('Test absolute correlation, no deviation', () => {
    const count = 100;
    const correlation = 1;
    const mean = 100;
    const deviation = Number.MIN_VALUE;
    let result;
    let stats;

    // Generate array with correlated pairs
    before(() => {
      result = generate(count, correlation, mean, deviation);
      stats = inspect(result);
    });

    describe('Should generate values with no distribution', () => {
      it('should generate an array of specified size', () => {
        assert.deepEqual(result.length, count);
      });

      it('should produce deviation of 0', () => {
        const { xDeviation, yDeviation } = stats;
        expect(xDeviation).equal(0);
        expect(yDeviation).equal(0);
      });

      it('should produce constant min/max', () => {
        const { xExtent, yExtent } = stats;
        expect(xExtent[0]).equal(mean);
        expect(xExtent[1]).equal(mean);
        expect(yExtent[0]).equal(mean);
        expect(yExtent[1]).equal(mean);
      });

      it('should produce constant mean', () => {
        const { xMean, yMean } = stats;
        expect(xMean).equal(mean);
        expect(yMean).equal(mean);
      });
    });
  });

  describe('Test correlation 0.7, mean 100, deviation 1', () => {
    const count = 10000;
    const correlation = 0.7;
    const mean = 100;
    const deviation = 1;
    let result;
    let stats;

    before(() => {
      result = generate(count, correlation, mean, deviation);
      stats = inspect(result);
    });

    describe('Should generate correlated pairs approximately conforming to spec', () => {
      it('should generate an array of specified size', () => {
        assert.deepEqual(result.length, count);
      });

      it('should produce correct pearson correlation coefficient approximation', () => {
        const { r } = stats;
        assert.deepEqual(true, approximatelyEqual(r, correlation, 0.03));
      });

      it('should produce correct deviation approximation', () => {
        const { xDeviation, yDeviation } = stats;
        assert.deepEqual(true, approximatelyEqual(xDeviation, deviation, 0.03));
        assert.deepEqual(true, approximatelyEqual(yDeviation, deviation, 0.03));
      });

      it('should produce correct min/max approximation', () => {
        const { xExtent, yExtent } = stats;
        assert.deepEqual(true, approximatelyEqual(xExtent[0], mean, 6));
        assert.deepEqual(true, approximatelyEqual(xExtent[1], mean, 6));
        assert.deepEqual(true, approximatelyEqual(yExtent[0], mean, 6));
        assert.deepEqual(true, approximatelyEqual(yExtent[1], mean, 6));
      });

      it('should produce correct mean approximation', () => {
        const { xMean, yMean } = stats;
        assert.deepEqual(true, approximatelyEqual(xMean, mean, 0.04));
        assert.deepEqual(true, approximatelyEqual(yMean, mean, 0.04));
      });
    });
  });

  describe('Test correlation 0.5, mean 1000, deviation 0.5', () => {
    const count = 10000;
    const correlation = 0.5;
    const mean = 1000;
    const deviation = 100;
    let result;
    let stats;

    before(() => {
      result = generate(count, correlation, mean, deviation);
      stats = inspect(result);
    });

    describe('Should generate correlated pairs approximately conforming to spec', () => {
      it('should generate an array of specified size', () => {
        assert.deepEqual(result.length, count);
      });

      it('should produce correct pearson correlation coefficient approximation', () => {
        const { r } = stats;
        assert.deepEqual(true, approximatelyEqual(r, correlation, 0.03));
      });

      it('should produce correct deviation approximation', () => {
        const { xDeviation, yDeviation } = stats;
        assert.deepEqual(true, approximatelyEqual(xDeviation, deviation, 2));
        assert.deepEqual(true, approximatelyEqual(yDeviation, deviation, 2));
      });

      it('should produce correct min/max approximation', () => {
        const { xExtent, yExtent } = stats;
        assert.deepEqual(true, approximatelyEqual(xExtent[0], mean, 600));
        assert.deepEqual(true, approximatelyEqual(xExtent[1], mean, 600));
        assert.deepEqual(true, approximatelyEqual(yExtent[0], mean, 600));
        assert.deepEqual(true, approximatelyEqual(yExtent[1], mean, 600));
      });

      it('should produce correct mean approximation', () => {
        const { xMean, yMean } = stats;
        assert.deepEqual(true, approximatelyEqual(xMean, mean, 4));
        assert.deepEqual(true, approximatelyEqual(yMean, mean, 4));
      });
    });
  });
});


src/index.js
/* eslint-disable camelcase, func-names, no-console, no-param-reassign, no-plusplus,
   no-shadow, object-curly-newline, wrap-iife */

/* eslint-disable max-len */
/**
 * array-correl
 * Version: 1.0.1
 * Purpose: Generates an array of correlated pairs of numbers, and inspects array with pair-wise numbers
 * Sources and inspiration:
 *   http://memory.psych.mun.ca/tech/js/correlation.shtml
 *   http://stevegardner.net/2012/06/11/javascript-code-to-calculate-the-pearson-correlation-coefficient/
 *   http://hongqinlab.blogspot.com/2013/11/how-to-generated-correlated-random.html
 *   https://math.stackexchange.com/questions/114982/how-can-we-generate-pairs-of-correlated-random-numbers
 * By Bo Ericsson
 */
/* eslint-enable max-len */
/* global d3 */

let d3Lib;

try {
  d3Lib = d3;
} catch (e) {
  d3Lib = require('d3-array');
}

// Constants
const DEFAULT_CORRELATION = 0.7;
const DEFAULT_MEAN = 0;
const DEFAULT_DEVIATION = 1;

(function (exports) {
  const version = '1.0.1';

  // Compute correlation coefficient
  function getPearsonCorrelation(x, y) {
    let shortestArrayLength = 0;

    if (x.length === y.length) {
      shortestArrayLength = x.length;
    } else if (x.length > y.length) {
      shortestArrayLength = y.length;
      const ignored = x.length - shortestArrayLength;
      console.error(`x has more items than y, the last ${ignored} item(s) will be ignored`);
    } else {
      shortestArrayLength = x.length;
      const ignored = x.length - shortestArrayLength;
      console.error(`x has more items than y, the last ${ignored} item(s) will be ignored`);
    }

    const xy = [];
    const x2 = [];
    const y2 = [];

    for (let i = 0; i < shortestArrayLength; i++) {
      xy.push(x[i] * y[i]);
      x2.push(x[i] * x[i]);
      y2.push(y[i] * y[i]);
    }

    let sum_x = 0;
    let sum_y = 0;
    let sum_xy = 0;
    let sum_x2 = 0;
    let sum_y2 = 0;

    for (let i = 0; i < shortestArrayLength; i++) {
      sum_x += x[i];
      sum_y += y[i];
      sum_xy += xy[i];
      sum_x2 += x2[i];
      sum_y2 += y2[i];
    }

    const step1 = (shortestArrayLength * sum_xy) - (sum_x * sum_y);
    const step2 = (shortestArrayLength * sum_x2) - (sum_x * sum_x);
    const step3 = (shortestArrayLength * sum_y2) - (sum_y * sum_y);
    const step4 = Math.sqrt(step2 * step3);
    const answer = step1 / step4;

    return answer;
  }

  // Generate normal distribution
  function randomNormal(mean = DEFAULT_MEAN, dev = DEFAULT_DEVIATION) {
    return function () {
      let a;
      let b;
      let r;

      do {
        a = 2.0 * Math.random() - 1.0;
        b = 2.0 * Math.random() - 1.0;
        r = a * a + b * b;
      } while (r >= 1.0);

      return dev * a * Math.sqrt((-2.0 * Math.log(r)) / r) + mean;
    };
  }

  // Generate array with pair-wise correlated numbers
  function generate(count, correlation, mean, deviation) {
    // Validate arguments
    if (count === undefined) {
      throw new ReferenceError('Undefined count argument');
    }

    [
      { name: 'count', param: count },
      { name: 'correlation', param: correlation },
      { name: 'mean', param: mean },
      { name: 'deviation', param: deviation },
    ].forEach((d) => {
      if (d.param !== undefined) {
        if (Number.isNaN(parseFloat(d.param)) || !Number.isFinite(d.param)) {
          throw new TypeError(`Invalid ${d.name} argument`);
        }
      }
    });

    // Initialize variables
    const r = correlation || DEFAULT_CORRELATION;
    const m = mean || DEFAULT_MEAN;
    const d = deviation || DEFAULT_DEVIATION;
    const nD = randomNormal();
    const data = [];

    // Generate correlated numbers
    d3Lib.range(count).forEach(() => {
      const xTemp = nD();
      const yTemp = nD();
      const x = m + d * xTemp;
      const y = m + d * (r * xTemp + (Math.sqrt(1 - r * r) * yTemp));
      data.push({ x, y });
    });

    // Return result
    return data;
  }

  // Analyze an array with pair-wise numbers
  function inspect(input) {
    // Validate argument
    if (!Array.isArray(input)) {
      throw new ReferenceError('Missing input array argument');
    } else if (Array.isArray(input[0]) && input[0].length !== 2) {
      throw new TypeError('Array length is not 2 at index 0 of input array');
    } else if (!Array.isArray() && !('x' in input[0] && 'y' in input[0])) {
      throw new TypeError('No x or y property at object at index 0 of input array');
    }

    // Get local copy of array
    let data = input.slice();

    // If input is in array-in-array format (not object-in-array), convert to object-in-array
    if (Array.isArray(input[0]) && data[0].length === 2) {
      data = [];
      for (let i = 0; i < input.length; i++) {
        data.push({
          x: input[i][0],
          y: input[i][1],
        });
      }
    }

    // Create x and y arrays
    const x = [];
    const y = [];
    data.forEach((d) => {
      x.push(d.x);
      y.push(d.y);
    });

    return {
      r: getPearsonCorrelation(x, y),
      xDeviation: d3Lib.deviation(data, (d) => d.x),
      yDeviation: d3Lib.deviation(data, (d) => d.y),
      xExtent: d3Lib.extent(data, (d) => d.x),
      yExtent: d3Lib.extent(data, (d) => d.y),
      xMean: d3Lib.mean(data, (d) => d.x),
      yMean: d3Lib.mean(data, (d) => d.y),
    };
  }

  // Exports
  exports.generate = (count, corr, mean, dev) => generate(count, corr, mean, dev);
  exports.inspect = (input) => inspect(input);
  exports.version = version;
})(typeof exports === 'undefined' ? this.arrayCorrel = {} : exports);


examples/node-example.js
/* eslint-disable no-console, no-plusplus */

// Import methods and version from array-correl from node_modules
// const { generate, inspect, version } = require('array-correl');

// Import methods and version from array-correl locally
const { generate, inspect, version } = require('../src/index.js');

// Log version
console.log(`array-correl version: ${version}`);

// Define inputs to the generate method
const count = 10000;
const correlation = 0.8;
const mean = 100;
const deviation = 1;

// Log inputs
console.log('\nInputs to generate:');
console.log(`  count: ${count}`);
console.log(`  correlation: ${correlation}`);
console.log(`  mean: ${mean}`);
console.log(`  deviation: ${deviation}`);

// Create an array with values of correlated pairs of numbers
const array = generate(count, correlation, mean, deviation);

console.log('\nFirst 10 correlated pairs:');
// Log the first 10 array elements
for (let i = 0; i < 10; i++) {
  console.log(`  ${JSON.stringify(array[i])}`);
}

// Inspect the array
const result = inspect(array);
console.log(`\nOutput of inspect: \n${JSON.stringify(result, null, 2)}\n`);

/*
Example output:
  array-correl version: 1.0.0

  Inputs to generate:
    count: 10000
    correlation: 0.8
    mean: 100
    deviation: 1

  First 10 correlated pairs:
    {"x":99.71068933565104,"y":99.86127640311815}
    {"x":100.83597240809056,"y":100.074669667149}
    {"x":102.0130498785201,"y":101.38201755974566}
    {"x":100.65061261210614,"y":99.88629765791357}
    {"x":98.30631250480867,"y":98.49421518032264}
    {"x":99.48427439034523,"y":99.81419151256067}
    {"x":98.90338120221003,"y":98.38761587336445}
    {"x":101.06519781042032,"y":101.50229105410696}
    {"x":98.7248721438115,"y":98.06148966509333}
    {"x":99.40834127473076,"y":98.90308575665651}

  Output of inspect:
  {
    "r": 0.802558017048035,
    "xDeviation": 1.0004768422899535,
    "yDeviation": 0.9984232776109369,
    "xExtent": [
      95.99318668690016,
      103.92406651194449
    ],
    "yExtent": [
      96.15152222172428,
      103.84111296300848
    ],
    "xMean": 100.00871360178544,
    "yMean": 100.00635903137874
  }
*/
